experimental/lib/Support/BooleanLogic/BoolExpression.cpp

```diff
@@ -0,0 +1,227 @@
+  SingleCond *singleCond = static_cast<SingleCond *>(this);
+  if (s[varIndex[singleCond->id]] == 'd')
+    s[varIndex[singleCond->id]] = singleCond->isNegated ? '0' : '1';
+  else if (s[varIndex[singleCond->id]] == '0')
+    s[varIndex[singleCond->id]] = singleCond->isNegated ? '0' : 'n';
+  else if (s[varIndex[singleCond->id]] == '1')
+    s[varIndex[singleCond->id]] = (!singleCond->isNegated) ? '1' : 'n';
+}
+
+void BoolExpression::generateMintermAnd(
+    std::string &s, const std::map<std::string, int> &varIndex) {
+  if (type == ExpressionType::Variable) {
+    generateMintermVariable(s, varIndex);
+  } else if (type == ExpressionType::Zero) { // 0 . exp =0 -> not a a minterm
+    s[0] = 'n';
+  } else if (type == ExpressionType::And || type == ExpressionType::Or) {
+    Operator *op = static_cast<Operator *>(this);
+    if (op->left != nullptr)
+      op->left->generateMintermAnd(s, varIndex);
+    if (op->right != nullptr)
+      op->right->generateMintermAnd(s, varIndex);
+  }
+}
+
+void BoolExpression::generateMintermsOr(
+    int numOfVariables, const std::map<std::string, int> &varIndex,
+    std::set<std::string> &minterms) {
+  if (type == ExpressionType::Or) {
+    Operator *op = static_cast<Operator *>(this);
+    op->left->generateMintermsOr(numOfVariables, varIndex, minterms);
+    op->right->generateMintermsOr(numOfVariables, varIndex, minterms);
+  } else if (type == ExpressionType::One) { // 1 + exp = 1;
+    std::string s(numOfVariables, 'd');
+    minterms.insert(s);
+  } else if (type == ExpressionType::Zero) { // 0 + exp = exp
+    return;
+  } else {
+    std::string s(numOfVariables, 'd'); // initializing s
+    generateMintermAnd(s, varIndex);
+    if (s.find('n') ==
+        std::string::npos) // no null in the minterm -> minterm is valid
+      minterms.insert(s);
+  }
+}
+
+// 1- generate all the minterms with don't cars
+// 2- replace the don't cares wit 0s and 1s
+// 3- loop over all rows in the truth table and check wether it corresponds to a
+// minterm or not
+std::set<std::string> BoolExpression::generateTruthTable() {
+  std::set<std::string> variables = BoolExpression::getVariables();
+  int numOfVariables = variables.size();
+  std::map<std::string, int> varIndex;
+  int index = 0;
+  for (const std::string &s : variables) {
+    varIndex[s] = index++;
+  }
+  // generate the minterms
+  std::set<std::string> minterms;
+  generateMintermsOr(numOfVariables, varIndex, minterms);
+  std::set<std::string> mintermsWithoutDontCares = replaceDontCares(minterms);
+  // generate the truth table
+  std::string s(numOfVariables, 'd');
+  std::set<std::string> r = {s};
+  std::set<std::string> rows = replaceDontCares(r);
+  std::set<std::string> table;
+  for (const std::string &row : rows) {
+    // if the row corresponds to the minterm, then f is 1
+    if (mintermsWithoutDontCares.find(row) != mintermsWithoutDontCares.end())
+      table.insert((row + "1"));
+    // else f is 0
```

Comment:
nit: I don't like the comment between the `if` and `else`, it somewhat breaks the connection between the two. Either remove it or put in the `else` block.